### PR TITLE
Added a step-by-step documentation for installing and building LPython.

### PR DESCRIPTION
**Pull request by kmr-srbh**
This PR adds the step-by-step installation instructions for setting up and building LPython on Linux, Windows, macOS and Windows with WSL. It overrides all of the dummy LFortran data previously present.

The steps for setting up the project on Windows with WSL are taken from #1501, courtesy of @faze-geek.

### CODE REVIEW

**Review by certik — APPROVED** [review]
I think this looks good. Thank you for improving the installation instructions!

File: doc/src/installation.md [hunk 1]
```diff
@@ -1,319 +1,199 @@
...
+        - Make sure you have `g++` and `cmake` installed. If not, install them using the below command:
+
+            ```bash
+            sudo apt-get install g++ cmake
+            ```
+
+        - Run the following command to install the build dependencies:
+
+            ```bash
+            sudo apt-get install binutils-dev build-essential zlib1g-dev bison re2c flex
```
Comment (Shaikh-Ubaid) [inline doc/src/installation.md:24]: Thanks for this @kmr-srbh! I have minor concern. I think `bison` and `re2c` can be installed in the conda environment, right? So do we still need these to be installed globally in the system?

Comment (Shaikh-Ubaid) [inline doc/src/installation.md:24]: Also, I think explicitly installing flex might not be needed on linux.

File: doc/src/installation.md [hunk 2]
```diff
@@ -1,319 +1,199 @@
...
+  For Windows, these are additional requirements:
+  - Miniforge Prompt
+  - Visual Studio (with "Desktop Development with C++" workload)
+
+- ### Set up your system
+    - Linux
+        - Make sure you have `g++` and `cmake` installed. If not, install them using the below command:
+
+            ```bash
+            sudo apt-get install g++ cmake
```
Comment (Shaikh-Ubaid) [inline doc/src/installation.md:18]: Since we are installing `build-essentials` below, I think g++ would not be needed to be explicitly installed.

**Comment by Shaikh-Ubaid** [comment]
Last time when I was setting up LFortran on Linux, I noted the extra-dependencies that were needed to be installed after setting up the conda environment:
1. `build-essentials` (for gcc, g++ and make)
2. `binutils-dev` (for error: Could not find BFD and also for stacktraces support)
3. `zstd-static=1.5.5` (for `zstd-static` not found during link)
4. `zlib1g-dev` (to fix `ld` error `-lz` not found)
5. `clang` (as it is used by lfortran for linking when compiling user source code)

I installed 1, 2, 4, and 5 using `sudo apt-get install`.
3 should be installed via conda.

Please note the above are for LFortran while setting up in Ubuntu/Linux operation system.

File: doc/src/installation.md (hunk 2 above)
Comment (kmr-srbh) [reply doc/src/installation.md:18]: You are right. I missed this as I had installed `g++` separately before building the project, getting the error, and installing `build-essential`.

**Comment by certik** [comment]
Yes. For everything else we should use the conda environment. Especially important for Bison, which only works with version 3.4 and often fails with newer versions.

File: doc/src/installation.md (hunk 1 above)
Comment (kmr-srbh) [reply doc/src/installation.md:24]: An explicit installation of `flex` is required for downloading other header files required for build.

Comment (kmr-srbh) [reply doc/src/installation.md:24]: You are right. I am making the changes.:smile::+1:

**Comment by Shaikh-Ubaid** [comment]
Also from https://lfortran.zulipchat.com/#narrow/stream/197339-general/topic/Building.20Lfortran/near/420682935, it seems `zlib1g-dev` should be installed via `conda`. So, summarizing the above, I think the steps to setup an LCompiler on Linux can be:

 
- Step 0 (Prerequisite): Download and install conda if not already installed.
- Step 1: Create the conda environment as per the appropriate environment file in the project
- Step2: `sudo apt-get install build-essentials binutils-dev clang` 

So, seems like just two steps.

Notes:
- we should update this environment files to also include `zstd-static` and `zlib1g-dev`
- we might be able to eliminate installing clang, but I think for now clang has been hard coded to be used for linking in LFortran. I think LFortran should ideally use the compiler specified by `LFortran_CC` for linking when compiling user source code.

File: doc/src/installation.md (hunk 1 above)
Comment (Shaikh-Ubaid) [reply doc/src/installation.md:24]: I think `flex` should already be installed in the `conda` environment. I guess you missed to create a conda environment?

**Comment by kmr-srbh** [comment]
> Also from https://lfortran.zulipchat.com/#narrow/stream/197339-general/topic/Building.20Lfortran/near/420682935, it seems `zlib1g-dev` should be installed via `conda`. So, summarizing the above, I think the steps to setup an LCompiler on Linux can be:
> 
>     * Step 0 (Prerequisite): Download and install conda if not already installed.
> 
>     * Step 1: Create the conda environment as per the appropriate environment file in the project
> 
>     * Step2: `sudo apt-get install build-essentials binutils-dev clang`
> 
> 
> So, seems like just two steps.
> 
> Notes:
> 
>     * we should update this environment files to also include `zstd-static` and `zlib1g-dev`
> 
>     * we might be able to eliminate installing clang, but I think for now clang has been hard coded to be used for linking in LFortran. I think LFortran should ideally use the compiler specified by `LFortran_CC` for linking when compiling user source code.

I am making the changes. :+1: Thank you very much for the help!🙂

**Comment by Shaikh-Ubaid** [comment]
>we should update this environment files to also include zstd-static and zlib1g-dev

It seems like `zlib1g-dev` is not present on conda-forge. So we need to be careful here.

@certik do you have any suggestions for installing `zlib1g-dev`. It seems it is used only for c/c++ zlib header files.

**Comment by kmr-srbh** [comment]
>     * we should update this environment files to also include `zstd-static` and `zlib1g-dev`

I am adding the packages. As @Shaikh-Ubaid pointed out, `zlib1g-dev` is not present on conda-forge. A global installation works. Is it fine?

**Comment by Shaikh-Ubaid** [comment]
>Is it fine?

Seem fine to me for now. We can update later as we figure out better approaches.

**Comment by Shaikh-Ubaid** [comment]
Also, I noticed that you are working in `main` branch directly (for example this PR or the commit above). It is usually suggested to use separate branches to submit developmental code. You use `main` branch to pull in latest changes from the remote repository. You use other branches to work on developmental/experimental code and later submit as a PR when ready.

**Comment by kmr-srbh** [comment]
You are right @Shaikh-Ubaid. To get started quickly, I had started work on the default branch. I have created new work-specific branches and am working on them now. 🙂

**Pull request merged**